### ISSUE TITLE
Panic in gfx-memory after many allocations

### ISSUE THREAD

**Issue by Dinnerbone** [gfx-rs/wgpu#989]
**Description**
After allocating many buffers, the application will panic with the following backtrace:
```
thread 'main' panicked at 'attempt to shift left with overflow', C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\allocator\general.rs:145:26
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:475
   1: core::panicking::panic_fmt
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\core\src\panicking.rs:85
   2: core::panicking::panic
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\core\src\panicking.rs:50
   3: gfx_memory::allocator::general::bit::BitSet::add
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\allocator\general.rs:145
   4: gfx_memory::allocator::general::GeneralAllocator<gfx_backend_vulkan::Backend>::alloc_from_entry<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\allocator\general.rs:504
   5: gfx_memory::allocator::general::GeneralAllocator<gfx_backend_vulkan::Backend>::alloc_block<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\allocator\general.rs:563
   6: gfx_memory::allocator::general::{{impl}}::alloc<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\allocator\general.rs:647
   7: gfx_memory::heaps::memory_type::MemoryType<gfx_backend_vulkan::Backend>::alloc<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\heaps\memory_type.rs:85
   8: gfx_memory::heaps::Heaps<gfx_backend_vulkan::Backend>::allocate_from<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\heaps\mod.rs:182
   9: gfx_memory::heaps::Heaps<gfx_backend_vulkan::Backend>::allocate<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-memory-0.2.1\src\heaps\mod.rs:145
  10: wgpu_core::device::Device<gfx_backend_vulkan::Backend>::create_buffer<gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\7ac706f\wgpu-core\src\device\mod.rs:480
  11: wgpu_core::hub::Global<wgpu_core::hub::IdentityManagerFactory>::device_create_buffer<wgpu_core::hub::IdentityManagerFactory,gfx_backend_vulkan::Backend>
             at C:\Users\dinne\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\7ac706f\wgpu-core\src\device\mod.rs:1022
  12: wgpu::backend::direct::{{impl}}::device_create_buffer
             at C:\Users\dinne\.cargo\git\checkouts\wgpu-rs-40ea39809c03c5d8\e3eadca\src\backend\direct.rs:894
  13: wgpu::Device::create_buffer
             at C:\Users\dinne\.cargo\git\checkouts\wgpu-rs-40ea39809c03c5d8\e3eadca\src\lib.rs:1518
  14: ruffle_render_wgpu::WgpuRenderBackend<ruffle_render_wgpu::target::TextureTarget>::register_shape_internal<ruffle_render_wgpu::target::TextureTarget>
             at W:\personal\ruffle\render\wgpu\src\lib.rs:258
  15: ruffle_render_wgpu::{{impl}}::register_shape<ruffle_render_wgpu::target::TextureTarget>
             at W:\personal\ruffle\render\wgpu\src\lib.rs:765
  16: ruffle_core::display_object::graphic::Graphic::from_swf_tag
             at W:\personal\ruffle\core\src\display_object\graphic.rs:23
  17: ruffle_core::display_object::movie_clip::MovieClipData::define_shape
             at W:\personal\ruffle\core\src\display_object\movie_clip.rs:2195
  18: ruffle_core::display_object::movie_clip::{{impl}}::preload::{{closure}}
             at W:\personal\ruffle\core\src\display_object\movie_clip.rs:268
  19: ruffle_core::tag_utils::decode_tags<slice<u8>*,closure-0>
             at W:\personal\ruffle\core\src\tag_utils.rs:304
  20: ruffle_core::display_object::movie_clip::MovieClip::preload
             at W:\personal\ruffle\core\src\display_object\movie_clip.rs:392
  21: ruffle_core::player::{{impl}}::preload::{{closure}}
             at W:\personal\ruffle\core\src\player.rs:791
  22: ruffle_core::player::{{impl}}::mutate_with_update_context::{{closure}}<closure-0,tuple<>>
             at W:\personal\ruffle\core\src\player.rs:1145
  23: ruffle_core::player::GcArena::mutate<closure-0,tuple<>>
             at C:\Users\dinne\.cargo\registry\src\github.com-1ecc6299db9ec823\gc-arena-0.2.0\src\arena.rs:161
  24: ruffle_core::player::Player::mutate_with_update_context<closure-0,tuple<>>
             at W:\personal\ruffle\core\src\player.rs:1092
  25: ruffle_core::player::Player::preload
             at W:\personal\ruffle\core\src\player.rs:788
  26: ruffle_core::player::Player::set_root_movie
             at W:\personal\ruffle\core\src\player.rs:409
  27: exporter::take_screenshot
             at .\src\main.rs:123
  28: exporter::capture_multiple_swfs
             at .\src\main.rs:297
  29: exporter::main
             at .\src\main.rs:400
  30: core::ops::function::FnOnce::call_once<fn() -> core::result::Result<tuple<>, alloc::boxed::Box<Error>>,tuple<>>
             at C:\Users\dinne\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ops\function.rs:227
```

This doesn't seem to happen in gfx-memory 0.2.0.

**Repro steps**
No standalone application this time sorry, but you can reproduce it in our project by:
- Cloning [ruffle](https://github.com/ruffle-rs/ruffle/)
- Updating gfx-memory (we have it locked to 0.2.0): `cargo update -p gfx-memory --precise 0.2.1`
- Download [this collection of swfs](https://github.com/gfx-rs/wgpu/files/5381371/newgrounds_sample.zip) and extract it somewhere
- Run: `cargo run --package exporter -- ./newgrounds_sample ./out`

You can use `-g` to switch graphics backends. I've reproduced on vulkan and dx12. Vulkan takes longer, dx12 does it after 2 swfs for me.

**Expected vs observed behavior**
I expect picnics, not panics.

**Extra materials**
Linked above. A trace would be huge, but I can make one if needed.

**Platform**
Seen on Windows 10 & Linux
Reproduced with Vulkan & DX12, not tried other backends.

**Comment by kvark** [comment]
Thank you for this beautiful issue!
The problem has been located, and the fix is coming in https://github.com/gfx-rs/gfx-extras/pull/35

**Comment by kvark** [comment]
Fixed in gfx-memory-0.2.2
